reaper/reaper_linux.go

```diff
@@ -0,0 +1,133 @@
+//go:build linux
+// +build linux
+
+package reaper
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/moby/sys/mountinfo"
+	"github.com/sirupsen/logrus"
+)
+
+// there is a kernel bug described in:
+// https://lore.kernel.org/all/YrShFXRLtRt6T%2Fj+@risky/ where fuse can cause a
+// pidns exit to deadlock. They symptoms are:
+//
+//    1. pidns pid 1 in S (sleeping) state
+//    2. that pid has zap_pid_ns_processes() in its /proc/pid/stack
+//    3. some fuse mount exists, and one of the threads from that fuse mount is
+//       stuck in fuse's request_wait_answer()
+//
+// if those conditions are true, we need to manually tear down the fuse
+// connection so the pidns can exit and docker doesn't get stuck. we can do
+// this by writing something to
+//
+// 	/sys/fs/fuse/connections/$dev_minor/abort
+//
+// where $dev_minor is the minor number from the fuse superblock mount.
+func checkIfFuseWedgedPidNs(pid int, taskID string) {
+	if !kernelStackHas(pid, "zap_pid_ns_processes") {
+		return
+	}
+	logrus.Debugf("found wedged pidns %d %s", pid, taskID)
+
+	// the kernel has already destroyed the mountinfo for the pidns' init,
+	// since it is pretty far along in do_exit(). we need to keep the tid
+	// of the fuse process around so we can inspect its mountinfo instead.
+	//
+	// ideally we'd use the pids cgroup (i.e. the docker API) here to
+	// figure out what tasks we should look at, but that *also* has been
+	// invalidated and is incorrect at this point. luckily for us the fuse
+	// daemon that's causing this hang in our production case is a child of
+	// init, so we can just look at that.
+	targetTid := 0
+	for _, tid := range childThreadsOfPid(pid) {
+		logrus.Debugf("checking %d for request_wait_answer", tid)
+		if kernelStackHas(tid, "request_wait_answer") {
+			targetTid = tid
+			break
```

Comment:
what's the chance of anyone doing something weird with multiple fuse daemons?

Reply:
I think we'll pick it up below if everything is in one mount namespace, since we look at all the mount infos (this is just figuring out which thread we can actually use, since the init pid's mountinfo has been freed already).

Containers could create their own mount namespace with multiple fuses, but I haven't seen that in practice with the lockups we're experiencing.